llvm_ir_dataset_utils/compile_time_analysis_tools/dataset_download/write_data_files.py

```diff
@@ -0,0 +1,125 @@
+respective tar files named [lang]_bc_files.tar. Each file added to a tar
+file is named bc_files/file[index].bc, where index is a number that is
+between the range of start_index and end_index (not including end) in
+"indices.csv" and is incremented by one (smallest index is 1).
+
+Example usage: python write_data_files.py /tmp
+"""
+
+import argparse
+from datasets import load_dataset, parallel
+from os import makedirs, listdir
+import threading
+import csv
+import tarfile
+from io import BytesIO
+from time import time
+
+
+def get_args():
+  """Function to return the provided storage argument for the script.
+
+  Returns: argparse.Namespace
+  """
+
+  parser = argparse.ArgumentParser(
+      description="Configure path to store bitcode files, and configure batch size."
+  )
+  parser.add_argument('storage', type=str, help='Path to the storage location.')
+  return parser.parse_args()
+
+
+def write_dataset_files_and_index_info(storage: str) -> None:
+  """Function to write each IR bitcode file to a tar archive.
+
+  The function first loads the ComPile dataset into a HF datasets
+  Dataset object. It does this using an experimental parallel backend to
+  slightly speed up load times. Then, a list of dictionaries is made,
+  where each dict contains the starting and ending index for each IR
+  file type based on language. For example, if the entire dataset
+  consisted of C and C++ IR modules, then the dictionary for C would
+  note language='c', starting_index=0, ending_index=(C++ starting
+  index). The +1 for the ending_index allows for direct use in range(a,
+  b) syntax, but is NOT suitable for right inclusive syntax. The entries
+  of each dictionary are then written to a CSV file name 'indices.csv'
+  for further use by other scripts. The contents of each dictionary are
+  used to provide information to the n number of threads, where n is the
+  number of languages in the dataset. Each thread when started calls the
+  create_tar() sub-function. The sub-function uses a generator
+  expression to access the bitcode files from a subset (taken by
+  language) of the original Dataset object. This allows for low memory
+  usage while performing in-memory writing of each bitcode file to a tar
+  archive which is named according to the given language (i.e.,
+  c_bc_files.tar).
+
+  Args:
+    storage: Storage location for the tar archives
+  """
+
+  def create_tar(dataset_subset, start_index: int, dir_name: str,
+                 language: str):
+    with tarfile.open(dir_name + '/' + language + '_bc_files.tar', 'a:') as tar:
+      for x in enumerate((dataset_subset[i]["content"]
+                          for i in range(0, dataset_subset.num_rows))):
+        tarinfo = tarfile.TarInfo(name=f'bc_files/file{x[0]+1+start_index}.bc')
+        file_obj = BytesIO(x[1])
+        tarinfo.size = file_obj.getbuffer().nbytes
+        tarinfo.mtime = time()
+        tar.addfile(tarinfo, fileobj=file_obj)
+
+  with parallel.parallel_backend('spark'):
```

Comment:
Can you add a comment on the performance benefits of using the parallel backend?